File: current/src/main/scala/eightyDays/scala211/bank/Bank.scala
package eightyDays.scala211.bank

import eightyDays.scala211.bank.account.{Account, Amount}
import eightyDays.scala211.bank.partner.{Identification, Partner}

import scala.util.Try


case class Bank(name: String, var partners: Map[Identification, Partner] = Map[Identification, Partner](), var accounts: Set[Account] = Set[Account]()) {
  def post(account: Account, action: => Try[Account]) = action.map{ updatedAccount =>
    accounts = accounts - account
    accounts = accounts + updatedAccount
    updatedAccount
  }

  def deposit(account: Account, value: Amount) = post(account, account.deposit(value))

  def withdraw(account: Account, value: Amount) = post(account, account.withdraw(value))

  def add(partnerId: Identification, accountFactory: (Partner => Account)): Account = {
    val newAccount = accountFactory(partners(partnerId))
    accounts = accounts + newAccount
    newAccount
  }

  def add(pPartner: Partner): (Identification, Bank) =
    partners
      .find(_._2 == pPartner)
      .map { entry => (entry._1, this) }
      .getOrElse {
        val result = Identification()
        (result, copy(partners = partners + (result -> pPartner)))
      }

  def find(predicate: Partner => Boolean): Option[Partner] = partners
    .find(p => predicate(p._2))
    .map(_._2)

  def filterPartners(predicate: Partner => Boolean): Set[Partner] =
    partners
      .filter(p => predicate(p._2))
      .values
      .toSet

  def filterAccounts(predicate: Account => Boolean): Set[Account] =
    accounts
      .filter(predicate)
}

File: master/main/java/eightyDays/java8/bank/Bank.java
package eightyDays.java8.bank;

import eightyDays.java8.bank.account.Account;
import eightyDays.java8.bank.partner.Identification;
import eightyDays.java8.bank.partner.Partner;

import java.math.BigDecimal;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Bank {
    private String name;
    private Map<Identification, Partner> partners = new HashMap<>();
    private Map<Identification, Account> accounts = new HashMap<>();

    public Bank(String pName) {
        name = pName;
    }

    public String getName() {
        return name;
    }

    public Set<Partner> getPartners() {
        return new HashSet<>(partners.values());
    }

    public Set<Account> getAccounts() {
        return new HashSet<>(accounts.values());
    }

    public void deposit(Identification identification, BigDecimal value) {
        accounts.get(identification).deposit(value);
    }

    public void withdraw(Identification identification, BigDecimal value) {
        accounts.get(identification).withdraw(value);
    }

    public Identification add(final Partner pPartner) {
        Identification result;
        if (partners.containsValue(pPartner)) {
            result = partners.entrySet().stream()
                    .filter(keyValue -> pPartner.equals(keyValue.getValue()))
                    .findFirst()
                    .get()
                    .getKey();
        } else {
            result = Identification.newIdentification();
            partners.put(result, pPartner);
        }
        return result;
    }

    public void add(Account account) {
        accounts.put(account.getNumber(), account);
    }

    public Optional<Partner> find(Predicate<Partner> predicate) {
        return partners.entrySet().stream()
                .map(Map.Entry::getValue)
                .filter( predicate )
                .findFirst();
    }

    public Set<Partner> filterPartners(Predicate<Partner> predicate) {
        return partners.entrySet().stream()
                .map(Map.Entry::getValue)
                .filter(predicate)
                .collect(Collectors.toSet());
    }

    public Set<Account> filterAccounts(Predicate<Account> predicate) {
        return accounts.entrySet().stream()
                .map(Map.Entry::getValue)
                .filter(predicate)
                .collect(Collectors.toSet());
    }
}
